File: 2023/nim/p06.nim
import strutils, sequtils, math

let
    input = readFile("../data/06.txt").strip.splitLines
    time = input[0].splitWhitespace[1..^1]
    distance = input[1].splitWhitespace[1..^1]

func f(t: uint64, d: uint64): int =
    let
        x = sqrt((t*t - 4*d).float)
        a = ((t.float-x)/2).floor.int + 1
        b = ((t.float+x)/2).ceil.int - 1
    b-a+1

# part 1
echo (0..3).mapIt(f(time[it].parseUInt, distance[it].parseUInt)).foldl(a*b)

# part 2
echo f(time.join.parseUInt, distance.join.parseUint)

File: 2023/ruby/p06.rb
input = IO.foreach('../data/06.txt').map(&:strip)

time = input[0].split[1..-1]
distance = input[1].split[1..-1]

def f(t, d)
  x = (t*t - 4*d)**0.5
  a = ((t-x)/2).floor + 1
  b = ((t+x)/2).ceil - 1
  b-a+1
end

# part 1
p 4.times.map{|i|
  f(time[i].to_i, distance[i].to_i)
}.reduce(:*)

# part 2
p f(time.join.to_i, distance.join.to_i)
